org/mozilla/jss/nss/SSL.c

```diff
@@ -365,6 +439,142 @@ Java_org_mozilla_jss_nss_SSL_SecurityStatus(JNIEnv *env, jclass clazz,
         issuer, subject);
 }
 
+JNIEXPORT jobject JNICALL
+Java_org_mozilla_jss_nss_SSL_GetChannelInfo(JNIEnv *env, jclass clazz,
+    jobject fd)
+{
+    PRFileDesc *real_fd = NULL;
+    SSLChannelInfo info = { 0 };
+    jint pV = 0;
+    jint cS = 0;
+    jint aKB = 0;
+    jint kKB = 0;
+    jlong cT = 0;
+    jlong lAT = 0;
+    jlong eT = 0;
+    jbyteArray sID = NULL;
+    jint cM = 0;
+    jboolean eMSU = JNI_FALSE;
+    jboolean eDA = JNI_FALSE;
+    jint kT = 0;
+    jint kG = 0;
+    jint sC = 0;
+    jint mA = 0;
+    jint aT = 0;
+    jint sS = 0;
+    jboolean haveNSS334 = JNI_FALSE;
+    jint oKG = 0;
+    jboolean r = JNI_FALSE;
+    jboolean pDC = JNI_FALSE;
+
+    PR_ASSERT(env != NULL && fd != NULL);
+    PR_SetError(0, 0);
+
+    if (JSS_PR_getPRFileDesc(env, fd, &real_fd) != PR_SUCCESS) {
+        return NULL;
+    }
+
+    if (SSL_GetChannelInfo(real_fd, &info, sizeof(info)) != SECSuccess) {
+        return NULL;
+    }
+
+    pV = info.protocolVersion;
+    cS = info.cipherSuite;
+
+    aKB = info.authKeyBits;
+    kKB = info.keaKeyBits;
+
+    cT = info.creationTime;
+    lAT = info.lastAccessTime;
+    eT = info.expirationTime;
+    sID = JSS_ToByteArray(env, info.sessionID, info.sessionIDLength);
+
+    cM = info.compressionMethod;
+
+    eMSU = info.extendedMasterSecretUsed;
+
+    eDA = info.earlyDataAccepted;
+
+    kT = info.keaType;
+    kG = info.keaGroup;
+    sC = info.symCipher;
+    mA = info.macAlgorithm;
+    aT = info.authType;
+    sS = info.signatureScheme;
+
+#ifdef HAVE_NSS_CHANNEL_INFO_ORIGINAL_KEA_GROUP
+    /* The following fields were added in NSS v3.34 and are detected
+     * via feature detection in CMake. */
+    haveNSS334 = JNI_TRUE;
```

Comment:
Is `haveNSS334` still needed? Or should it be renamed to something like `haveNSSChannelInfoOriginalKeaGroup` since the actual NSS version number is no longer relevant?

Reply:
I think the name is fine as-is. It is supposed to be present in that version or later. Also, in the `SSLPreliminaryChannelInfo`, it would conflict with the query field's implementation:

https://github.com/dogtagpki/jss/blob/eeeeaa5513dc8521099a08f08c9a8b04dbb4ffd0/org/mozilla/jss/nss/SSLPreliminaryChannelInfo.java#L219-L228

So I'd perfer to leave it as-is, to show the separation.

Reply:
Since the javadoc comments describe what the parameters represent, I guess this is fine for now. In the long term the more we have these discrepancies between version-based and feature-based checks the code could become more difficult to understand.

Reply:
Okie dokie, thanks! (And yeah, I'm hoping most version/feature discrepancies can be hidden in the future and kept at the C level like the SSLCipher ones. ).